fedbiomed/common/utils/_secagg_utils.py

```diff
@@ -1,6 +1,38 @@
 # This file is originally part of Fed-BioMed
 # SPDX-License-Identifier: Apache-2.0
 
+def matching_parties_dh(context: dict, parties: list) -> bool:
+    """Check if parties of given context are compatible with the parties
+        of a secagg Diffie Hellman element.
+
+        Args:
+            context: context to be compared with the secagg servkey element parties
+            parties: the secagg servkey element parties
+
+        Returns:
+            True if this context can be used with this element, False if not.
+        """
+    # Need to ensure that:
+    # - first party needs to be the same for both (checked for consistency, but not security important)
+    # - existing element was established for the same parties or a superset of the parties
+    #   (order can differ, as nodes are ordered by the cipher code)
+    #
+    # eg: [ 'un', 'deux', 'trois' ] parties compatible with [ 'un', 'trois', 'deux' ] context
+    # but not with [ 'deux', 'un', 'trois' ]
+    # eg: [ 'un', 'deux', 'trois' ] parties compatible with [ 'un', 'trois', 'quatre', 'deux' ] context
+    # but not with [ 'un', 'deux', 'quatre' ]
+    return (
+        # Commented tests can be assumed from calling functions
+        #
+        # isinstance(context, dict) and
+        # 'parties' in context and
+        # isinstance(context['parties'], list) and
+        # len(context['parties']) >= 1 and
+        # isinstance(parties, list) and
+        parties[0] == context['parties'][0] and
```

Comment:
Hi @mvesin,

I didn't understand why the first party need to be same?

Reply:
Hi @srcansiz 

> checked for consistency, but not security important

If I understand it correctly, this is not really needed in the sense that there is no cryptographic material on the researcher (`parties[0]`) , so technically a `secagg_id` could be used by another researcher.

But on the other hand, a `secagg_id` is attached to an `experiment_id`, and an experiment is itself attached to a `researcher_id`. So it should not be used with another researcher, when we implement multiple researchers in the future.
So I would say this is good practice to implement that already, though I agree it does not change anything for now, as we use single `researcher_id`

Reply:
Ok I see you point. However, researcher is not part of this secagg context. Therefore, I would not include researcher in the list of parties. In terms of consistency, I think the consistency between experiment and researcher should be done before the code reaches this point. Node should verify if the experiment belongs to that researcher or not before processing anything (maybe just after receiving a request related to a experiment). However, we do not have this mechanism yet so I would leave it to later rather than implementing now.

Reply:
OK, then I can remove the test for `parties[0] == context['parties'][0]`.

Nevertheless, I would like to keep passing *all* parties as arguments, to have common prototype with the other `matching_parties_*()` functions. So we can easier refactor this code later.
Servkey and biprime need all the parties to properly check compatibility of parties with the secagg context element.